File: Search and Filter/log-api-server/alertEngine.js
function checkAlerts(logs) {
  const alerts = [];

  const now = new Date();
  const fiveMinutesAgo = new Date(now.getTime() - 5 * 60 * 1000);

  // Rule: ERROR count > 5 in last 5 minutes
  const recentErrors = logs.filter(log => {
    if (!log.level) return false;
    const logTime = new Date(log.time.replace(" ", "T"));
    return log.level === "ERROR" && logTime >= fiveMinutesAgo;
  });

  if (recentErrors.length > 5) {
    alerts.push({
      name: "High Error Rate",
      severity: "HIGH",
      reason: `More than 5 ERROR logs in last 5 minutes`,
      stats: {
        count: recentErrors.length,
        timeWindow: "5 minutes"
      }
    });
  }

  // Rule: Keyword 'timeout' appears > 3 times
  const keyword = "timeout";
  const keywordMatches = logs.filter(log =>
    log.message &&
    log.message.toLowerCase().includes(keyword)
  );

  if (keywordMatches.length > 3) {
    alerts.push({
      name: "Timeout Spike",
      severity: "LOW",
      reason: `'${keyword}' appeared more than 3 times`,
      stats: {
        count: keywordMatches.length,
        keyword
      }
    });
  }

  return alerts;
}

module.exports = { checkAlerts };

File: Search and Filter/log-api-server/alert_engine.py
# alert_engine.py

def get_alerts(df):
    alerts = []

    err_count = len(df[df['level'] == 'ERROR'])
    if err_count > 10:
        alerts.append({
            "name": "Critical Error Spike",
            "severity": "HIGH",
            "reason": f"System has {err_count} errors."
        })

    keyword_count = len(df[df['message'].str.contains("Asset in use", case=False)])
    if keyword_count > 5:
        alerts.append({
            "name": "Resource Contention",
            "severity": "LOW",
            "reason": f"'Asset in use' seen {keyword_count} times."
        })

    return alerts
